### PR TITLE
ChibiOS: added monitor of MCU temperature and voltage

### PR DESCRIPTION
**Pull request by tridge**
This allows us to track the internal temperature and voltage of the MCU. The voltage is given as an average/min/max so we have an idea of how stable the power supply is
This needs https://github.com/ArduPilot/mavlink/pull/224

I'm seeing about a 0.02V variation in bench testing on USB power:
image

### CODE REVIEW

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp [hunk 1]
```diff
@@ -289,6 +293,119 @@ void AnalogIn::read_adc(uint32_t *val)
...
+    }
+    sample_adc3_count += ADC_DMA_BUF_DEPTH;
+}
+
+/*
+  setup ADC3 for internal temperature and voltage monitoring
+ */
+void AnalogIn::setup_adc3(void)
+{
+    samples_adc3 = (adcsample_t *)hal.util->malloc_type(sizeof(adcsample_t)*ADC_DMA_BUF_DEPTH*ADC3_GRP1_NUM_CHANNELS, AP_HAL::Util::MEM_DMA_SAFE);
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/AnalogIn.cpp:347]: not checking if allocation was successful?

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp [hunk 2]
```diff
@@ -289,6 +293,119 @@ void AnalogIn::read_adc(uint32_t *val)
...
+uint16_t AnalogIn::sample_adc3_max[ADC3_GRP1_NUM_CHANNELS];
+uint16_t AnalogIn::sample_adc3_min[ADC3_GRP1_NUM_CHANNELS];
+uint32_t AnalogIn::sample_adc3_count;
+
+/*
+  callback from ADC3 driver when sample buffer is filled
+ */
+void AnalogIn::adc3callback(ADCDriver *adcp)
+{
+    const adcsample_t *buffer = samples_adc3;
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/AnalogIn.cpp:322]: should check and return if null

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp [hunk 3]
```diff
@@ -289,6 +293,119 @@ void AnalogIn::read_adc(uint32_t *val)
...
+#define ADC3_VSENSE_CHAN 18
+#define ADC3_VREFINT_CHAN 19
+#define ADC3_VBAT4_CHAN 17
+
+// samples filled in by ADC DMA engine
+adcsample_t *AnalogIn::samples_adc3;
+uint32_t AnalogIn::sample_adc3_sum[ADC3_GRP1_NUM_CHANNELS];
+// we also keep min and max so we can report the range of voltages
+// seen, to give an idea of supply stability
+uint16_t AnalogIn::sample_adc3_max[ADC3_GRP1_NUM_CHANNELS];
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/AnalogIn.cpp:313]: should be `adcsample_t`

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp [hunk 4]
```diff
@@ -349,19 +466,28 @@ void AnalogIn::_timer_tick(void)
         }
     }
 
-#if CHIBIOS_ADC_MAVLINK_DEBUG
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/AnalogIn.cpp:352]: better to remove it from the places it is used as well to remove confusion. Like start of this file and also in skyviper hwdefs.

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py [hunk 5]
```diff
@@ -233,9 +233,9 @@ def generate_DMAMUX_map(peripheral_list, noshare_list, dma_exclude, stream_ofs):
     map1 = generate_DMAMUX_map_mask(dmamux1_peripherals, 0xFFFF, noshare_list, dma_exclude, stream_ofs)
     # there are 8 BDMA channels, but an issue has been found where if I2C4 and SPI6
     # use neighboring channels then we sometimes lose a BDMA completion interrupt. To
-    # avoid this we set the BDMA available mask to 0x33, which forces the channels not to be
+    # avoid this we set the BDMA available mask to 0x54, which forces the channels not to be
     # adjacent. This issue was found on a CUAV-X7, with H743 RevV.
-    map2 = generate_DMAMUX_map_mask(dmamux2_peripherals, 0x55, noshare_list, dma_exclude, stream_ofs)
+    map2 = generate_DMAMUX_map_mask(dmamux2_peripherals, 0x54, noshare_list, dma_exclude, stream_ofs)
```
Comment (bugobliterator) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py:238]: does this change previous mapping results?

**Review by bugobliterator — CHANGES_REQUESTED** [review]
Generally looks good to me, except for few minor issues and queries

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/AnalogIn.cpp:344]: fixed to not init adc3 if alloc fails

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/AnalogIn.cpp:319]: as it now skips the init if alloc fails then the callback is never setup

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/AnalogIn.cpp:310]: i've added a static_assert that adcsample_t and uint16_t are the same size

File: libraries/AP_HAL_ChibiOS/AnalogIn.cpp (hunk 4 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/AnalogIn.cpp:352]: done

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py (hunk 5 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py]: this one is complex (good EU dev call topic)
I did some more testing and found that both ADC3 and SPI6_RX can't use stream 0. On top of the existing known issues with BDMA I decided to just statically map the 3 peripherals that can use BDMA to fixed streams.
I also fixed SPI6 to stop trying to do BDMA locking, as we now can never have shared BDMA streams

**Review by bugobliterator — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py (hunk 5 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/dma_resolver.py]: yeah, the change to using statically assigned channels make sense.

**Pull request merged**